rfd/0074-kube-secret-storage.md

```diff
@@ -0,0 +1,603 @@
+    app.kubernetes.io/instance: {{ .Release.Name | quote }}
+    app.kubernetes.io/version: {{ .Chart.AppVersion }}
+    helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+  annotations:
+    "helm.sh/hook": post-upgrade
+    "helm.sh/hook-weight": "-5"
+    "helm.sh/hook-delete-policy": hook-succeeded
+spec:
+  template:
+    metadata:
+      name: "{{ .Release.Name }}"
+      labels:
+        app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
+        app.kubernetes.io/instance: {{ .Release.Name | quote }}
+        helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: post-upgrade-job
+        image: "alpine:3.3"
+        command: ["/bin/bash","-c", "kubectl rollout status --watch --timeout=600s statefulset/{{ .Release.Name }} && kubectl delete deployment/{ .Release.Name } -n { .Release.Namespace }"]
++{{- end}}
+```
+
+
+## UX
+
+If the agent detects that it is running inside Kubernetes, it will enable, by default, the Kube secret storage instead of local storage. This means that it will not require any additional configuration or operation by the end-user.
+
+To install the agent on Kubernetes using Helm, the command remains the same as before, as one can see below.
+
+```bash
+$ helm install/upgrade teleport-agent  teleport/teleport-kube-agent \
+  --create-namespace \
+  --namespace teleport-agent \
+  --set roles=kube \
+  --set proxyAddr=${PROXY_ENDPOINT?} \
+  --set authToken=${JOIN_TOKEN?} \
+  --set kubeClusterName=${KUBERNETES_CLUSTER_NAME?}
+```
+
+## Security
+
+The change with the greatest impact is that the agent's service account has read and write permissions for any existing secret in the namespace where it was installed. If the SA token is stolen or a hijack happens in the Teleport agent, the attacker can use it to read any secrets in that namespace. We recommend that Teleport is installed in a separate namespace to minimize this risk.
+
+This issue happens because, currently, Kubernetes does not allow the specification of `resourceNames` in RBAC roles with [wildcards][wildcards]. If supported in the future, we can limit the read/list/update actions when handling secrets for the Teleport Service account. Otherwise, another possible solution might be to keep the identity stored in a single Secret instead of a secret per pod, allowing read and write operations only for that secret.
+
+Another possible security concern is Secret data encryption. By default, standard Kubernetes clusters store objects in `etcd` as plaintext, but most cloud vendors ship Kubernetes Clusters with encryption at REST for objects, including Secrets. Although they offer such a solution, it is possible to add application-level encryption using a KMS for managing sensitive data such as Secrets. Application-layer secret encryption provides an additional layer of security for sensitive data since it protects against attackers who gain access to an offline copy of etcd.
```

Comment:
Since as pointed out k8s secrets are not encrypted at rest by default (even if [support for encryption](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#configuration-and-determining-whether-encryption-at-rest-is-already-enabled) is available from Kubernetes 1.7), this should be pointed out in the Teleport docs. Other vendors also addressed this using volume-level encryption (for example, dm-crypt or cloud provider volume encryption) for the etcd data volumes (but ofc this is not always possible).